File: 2013/20130220 - caixas - python/caixas.py
#-*- coding: utf-8 -*-

class Caixa(object):
	def __init__(self, peso, capacidade):
		self.peso = peso
		self.capacidade = capacidade

INFINITO = 1e300

def empilhar(caixas):
	caixas_sort_capacidade = caixas.sort(key=lambda caixa: (-caixa.capacidade, -caixa.peso))

	capacidade_da_pilha = INFINITO
	tamanho_da_pilha = 0

	for caixa in caixas_sort_capacidade:
		if caixa.peso <= capacidade_da_pilha:
			capacidade_da_pilha = min(
				capacidade_da_pilha - caixa.peso, caixa.capacidade)
			tamanho_da_pilha += 1

	return tamanho_da_pilha

File: 2013/20130522 - caixas - ruby/caixas.rb
class Caixas

  INFINITO = 1.0/0

  def self.pode?(em_cima, capacidade)
    em_cima[:peso] <= capacidade
  end

  def self.empilhar(caixas, capacidade = INFINITO)
    caixas = caixas.sort_by { |e| -e[:capacidade] / e[:peso].to_f}

    pilha = 0
    capacidade_total = INFINITO

    caixas.each do |caixa|
      if pode?(caixa, capacidade_total)
        capacidade_total = [capacidade_total - caixa[:peso], caixa[:capacidade]].min
        pilha += 1
      end
    end

    pilha
  end
end
